### PR TITLE
[editor] feat: copy/paste from google docs & ms word with formatting

### PR DESCRIPTION
**Pull request by maimonav**
Copy/Paste functionalities:

- Support inline styling (line spacing, bold, italic, underline) on lists & headers (all other elements were already supported).

- Fix all bold content from google docs.

- Normalize ms word html.

- Support headers.

- Support alignments.

- Unit tests added.

### CODE REVIEW

File: packages/editor/web/src/RichContentEditor/handlePastedText.js [hunk 1]
```diff
@@ -25,29 +26,48 @@ const clearAtomicBlockEntities = editorState => {
 
 const FRAGMENT_ATTR = 'data-draftjs-conductor-fragment';
 
-//Fixes spaces deletion when pasting hyperlinks (draft-convert trimming)
-const replaceSpansWithOnlySpaces = html =>
-  // eslint-disable-next-line no-irregular-whitespace
-  html.replace(/(<span> <\/span>)/g, ' ');
+//Fix replaceWithFragment when fragment size == 1, https://github.com/facebook/draft-js/issues/1511
+const replaceWithFragment = (editorState, contentState, contentToPaste) => {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/handlePastedText.js:30]: check if it works with the original signature

File: packages/editor/web/src/RichContentEditor/handlePastedText.js [hunk 2]
```diff
@@ -25,29 +26,48 @@ const clearAtomicBlockEntities = editorState => {
...
+  let contentWithPaste = Modifier.replaceWithFragment(
+    contentState,
+    editorState.getSelection(),
+    contentToPaste.getBlockMap()
+  );
+
+  const shouldOverrideBlockType = blockType => blockType === 'unstyled' || blockType === 'atomic';
```
[the remark this answers is not in the document]
Comment (shaulgo) [reply packages/editor/web/src/RichContentEditor/handlePastedText.js:37]: override always and only when pasted to block is empty

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  h1: 'header-one',
+  h2: 'header-two',
+  h3: 'header-three',
+  h4: 'header-four',
+  h5: 'header-five',
+  h6: 'header-six',
+};
+
+const getDraftTypeOfElement = element => {
+  const type = headerElementToDraftType[element] ?? element;
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:14]:
```suggestion
  const type = headerElementToDraftType[element] || 'unstyled';
```

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+const getDynamicStyles = style => {
+  if (!style?.lineHeight && !style?.paddingTop && !style?.paddingTop) {
+    return undefined;
+  }
+
+  const dynamicStyles = {
+    'line-height': style?.lineHeight,
+    'padding-top': style?.paddingTop,
+    'padding-bottom': style?.paddingTop,
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:26]:
```suggestion
    'padding-bottom': style?.paddingBottom,
```

File: packages/editor/web/src/RichContentEditor/utils/normalizeHTML.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  if (match && match[1]) {
+    htmlString =
+      htmlString.slice(0, match.index) + htmlString.slice(match.index).replace(match[1], '');
+  }
+
+  return htmlString;
+};
+
+export default function normalizeHTML(htmlString) {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/normalizeHTML.js:60]: unit test

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  type === 'li' ||
+  type === 'p' ||
+  type === 'h1' ||
+  type === 'h2' ||
+  type === 'h3' ||
+  type === 'h4' ||
+  type === 'h5' ||
+  type === 'h6';
+
+export default function htmlToBlock(nodeName, node) {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:46]: unit test

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    'line-height': style?.lineHeight,
+    'padding-top': style?.paddingTop,
+    'padding-bottom': style?.paddingTop,
+  };
+  return pickBy(dynamicStyles);
+};
+
+const getTextAlignment = style => {
+  const nodeTextAlign = style?.textAlign;
+  return !!nodeTextAlign && nodeTextAlign !== 'start' ? nodeTextAlign : 'left';
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:33]: recheck this. If no textAlign, then no need to add alignment.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+const isElementToBlock = type =>
+  type === 'li' ||
+  type === 'p' ||
+  type === 'h1' ||
+  type === 'h2' ||
+  type === 'h3' ||
+  type === 'h4' ||
+  type === 'h5' ||
+  type === 'h6';
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:44]:
```suggestion
['li', 'p'].includes(type)
```

Comment (shaulgo) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:44]: whatever you prefer...

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+export default function htmlToBlock(nodeName, node) {
+  let type, style;
+
+  if (isElementToBlock(nodeName)) {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:49]:
```suggestion
  if (shouldConvertElementToBlock(nodeName)) {
```

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    if (nodeName === 'li') {
+      if (node.parentElement.nodeName === 'OL') {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:51]: maybe use ?

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        type = 'ordered-list-item';
+      } else {
+        type = 'unordered-list-item';
+      }
+      style = node.firstChild.style || node.style;
+    } else {
+      type = getDraftTypeOfElement(nodeName);
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:58]:
```suggestion
      type = getBlockTypeOfElement(nodeName);
```

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: packages/editor/web/src/RichContentEditor/handlePastedText.js (hunk 1 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/handlePastedText.js:30]: done.

File: packages/editor/web/src/RichContentEditor/handlePastedText.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/handlePastedText.js:37]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 3 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:14]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 4 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:26]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 8 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:44]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 10 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:51]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 11 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:58]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 9 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:49]: done.

File: packages/editor/web/src/RichContentEditor/RichContentEditor.tsx [hunk 12]
```diff
@@ -741,7 +741,11 @@ class RichContentEditor extends Component<RichContentEditorProps, State> {
     <AccessibilityListener isMobile={this.contextualData.isMobile} />
   );
 
-  styleToClass = ([key, val]) => `rich_content_${key}-${val.toString().replace('.', '_')}`;
+  styleToClass = ([key, val]) =>
+    `rich_content_${key}-${val
+      .toString()
+      .replace('.', '_')
+      .replace('%', '_percent')}`;
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/RichContentEditor.tsx:748]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 7 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:33]: done.

File: packages/editor/web/src/RichContentEditor/utils/normalizeHTML.js (hunk 5 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/normalizeHTML.js:60]: done.

File: packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js (hunk 6 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/utils/htmlToBlock.js:46]: Covered by normalizeHTML tests & handlePastedText tests.

File: packages/editor/web/src/RichContentEditor/handlePastedText.js [hunk 13]
```diff
@@ -25,29 +26,51 @@ const clearAtomicBlockEntities = editorState => {
...
+    ? draftConvertFromHtml(pastedContentConfig)(html)
     : ContentState.createFromText(text);
 
   const contentState = clearAtomicBlockEntities(editorState);
-  const contentWithPaste = Modifier.replaceWithFragment(
+  const contentWithPaste = replaceWithFragment(
     contentState,
     editorState.getSelection(),
-    contentToPaste.getBlockMap()
+    contentToPaste
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/handlePastedText.js:65]: let's keep the same signature.

Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/handlePastedText.js:65]: Sure.

**Review by shaulgo — APPROVED** [review]

**Comment by shaulgo** [comment]
add changelog

**Pull request merged**